Exporting sta data

mahjahromi opened the pull request:
[JIRA](https://sensorup.atlassian.net/browse/SEN-###)

## Summary

A download button is added to each sensor card (in components/sensor/sensor-item.vue) which would direct the user to the /sensor/download/“sensor MAC ID” page.

In the download page:
* The sensor’s MAC ID can be seen both in the url (ex: http://localhost:8080/sensors/download/24ED33) and the title of the page (ex: Device 24ED33 / Export)
* There is a filter section where user can select his desired Observed Properties (default values are Humidity, Temperature and PM2.5) and set the date rang
* In the Export section user can see a preview of the subset of the data using “Show preview” button, or download the CSV file using “Export CSV” button
* By pressing the “Export CSV” button a preview of the dataset is also shown on the page
* For the preview option the number of requested entities are limited to top 10, and that of for the download option is set to top 2000
* The CSV file contains following headers:
  - Observation Phenomenon Time (UTC) : 2019-08-02T15:45:45.909Z
  - Observation Result: 45.37
  - Location Name: Ahuntsic
  - Location Description: Montreal
  - Location GeoJSON: {"coordinates":[-73.69282,45.5451],"type":"Point"}
  - Datastream Unit of Measurement Symbol: %
  - Observed Property Description: Humidity
  - Sensor Name: SHT31
* For each of the preview or download option, there is one query for getting the Datastream ID(s) of the selected Observed Properties, and one query for each of the Observed Properties. 

## Notes
A Vue-warn is logged in the console: The computed property "fields" is already defined as a prop. Referring to [this issue](https://github.com/jecovier/vue-json-excel/issues/48) of vue-json-excel, the solution is configuring vee-validate to use a different name for the fieldsbag.

## Screenshots
download Icon added to each sensor card (It is only displayed when mouse is hover sensor card)
<img width="1440" alt="Screen Shot 2019-08-02 at 3 06 39 PM" src="https://user-images.githubusercontent.com/36073808/62398843-2465f080-b537-11e9-8801-c0396f9a9a90.png">

Download page for device 24ED33
<img width="1440" alt="Screen Shot 2019-08-02 at 3 07 00 PM" src="https://user-images.githubusercontent.com/36073808/62398855-3051b280-b537-11e9-8598-740dabd7d66d.png">

Observed Properties multi select options
<img width="1440" alt="Screen Shot 2019-08-02 at 3 08 46 PM" src="https://user-images.githubusercontent.com/36073808/62398949-70189a00-b537-11e9-9979-cd29ed4b810c.png">

Date range
<img width="1440" alt="Screen Shot 2019-08-02 at 3 09 13 PM" src="https://user-images.githubusercontent.com/36073808/62398964-7f97e300-b537-11e9-8e96-e967c016be62.png">

"Show Preview" button displays a preview of data (Top 10)
<img width="1440" alt="Screen Shot 2019-08-02 at 3 10 07 PM" src="https://user-images.githubusercontent.com/36073808/62399000-9fc7a200-b537-11e9-8993-c6bf68603b61.png">

Export CSV button is disabled and a spinner is shown during getting data from server
<img width="1440" alt="Screen Shot 2019-08-02 at 3 10 53 PM" src="https://user-images.githubusercontent.com/36073808/62399043-ba9a1680-b537-11e9-8910-e45672a03b30.png">

## Where to start
1. From the senor page select the download button of a sensor card
2. Try different buttons in the following page

mahjahromi committed:
download route and button added

mahjahromi committed:
Create readMe.md

mahjahromi committed:
Merge branch 'Exporting-STA-data' of https://github.com/canarie-geocens/smart-air-user-portal into Exporting-STA-data

mahjahromi committed:
sta headers and staDataExporter added

mahjahromi committed:
sta request and download added

mahjahromi committed:
deleted readme

mahjahromi committed:
installed dependencies

mahjahromi committed:
Merge branch 'Exporting-STA-data' of https://github.com/canarie-geocens/smart-air-user-portal into Exporting-STA-data

mahjahromi committed:
flat picker dependencies added

mahjahromi committed:
improving staDataExporter

mahjahromi committed:
send parameters through sensor-item

mahjahromi committed:
Update download.vue

mahjahromi committed:
Update sta-request.vue

mahjahromi committed:
dependency added

mahnoooosh commented:
Additionally, in the download page in case of server response of codes 400, 404, 500, 503, a dialog box will pop up showing response status.

kanluo commented:
Overall, this looks good to me. Good job. I find some problems you could fix. 

**Important**
- It didn’t display preview data correctly when you provide a given data range. It either displays the latest 10 data or nothing. 

- There are several unnecessary `=` in the exported data line. Is this on purpose?  If so, why? If not, please remove them.
```
"=""2019-08-05T18:04:12.952Z""","=""32.57""","=""Triangle Mtn - Langford""","=""Victoria""","=""{""coordinates"":[-123.514672,48.418927],""type"":""Point""}""","=""%""","=""Humidity""","=""SHT31 sensor"""
```
**Optional**
- Could put property description and unit next to observation result (result, property, unit)
- Might make the preview table higher to disable the table inline scroll.  Therefore, the user can scroll down the page with only the page scroll, which is more easily. 
- Might add a link,  which is a go back to the sensor list link,  next to the Device name, like [Sensor List](previous_link) / Device 24EC64 / Export
- Remove all selected properties from the dropdown list to reduce confusion, in other words, the dropdown list only display unselected properties
- Disabled the ‘show preview’ button during the data exporting as well

kanluo commented on src/components/data-exporter/sta-headers.js:
```diff
@@ -0,0 +1,155 @@
+  .value()
+
+export default {
+  headers: headers,
```
Where do you use the `headers`?

soroushojagh6633 replied:
We are using headers in download.vue file in order to fill in the table for preview and download returned data from the server. To be more precise, different entities have their own different properties, so to reduce the complexity we defined all the required entities and their properties in the sta-header javascript file. For example, we have different 'description' fields for different entities such as Things, Location, and Datastreams.

soroushojagh6633 replied:
We because we used it in the 'table', exporting it was redundant. Thanks, @kanluo for your suggestion.

kanluo replied:
Yeah. I understand your idea but I didn't find out where you used this. It doesn't matter even if you delete headers from the exporting like this:
```
export default {
  table,
}
```

soroushojagh6633 replied:
Yes, I got it was redundant and I delete it. Thanks @kanluo

kanluo commented on src/components/data-exporter/staDataExporter.vue:
```diff
@@ -0,0 +1,80 @@
+  },
+  computed: {
+    fileName() {
+      return this.name + '.' + this.type
```
could be `${this.name}.${this.type}`.

kanluo commented on src/router/views/download.vue:
```diff
@@ -0,0 +1,313 @@
+    import MainContent from '@layouts/main-content';
+    import staRequests from './sta-request';
+    import HotTable from '@handsontable/vue';
+    import staHeaders from '@components/data-exporter/sta-headers.js';
```
Might change to import {table} from '@components/data-exporter/sta-headers.js'; and replace `staHeaders.table` to `table` directly

kanluo commented on src/router/views/sta-request.js:
```diff
@@ -0,0 +1,280 @@
+    requiredSensors: false,
+    requiredObservedProperties: false,
+
+    async getSpecificThing(thingsId, obsrvedProperttyArray, timeRange, selectedFilters,option) {
```
This function is too long.  Might be split into several sub-functions

soroushojagh6633 replied:
Thanks for your feedback, it is done in the new version @kanluo

mahnoooosh commented:
> Overall, this looks good to me. Good job. I find some problems you could fix.
> 
> **Important**
> 
> * It didn’t display preview data correctly when you provide a given data range. It either displays the latest 10 data or nothing.

**Just in order to make sure I understand what you've experienced, the display button is supposed to show top 10 data and your selected date range is considered in the query as:
https://canada-geocens-aq-sta.sensorup.com/v1.0//Datastreams(53444042)/Observations?$count=false&$filter=Datastreams%2FObservations%2FphenomenonTime+ge+2019-08-01T06:00:00.000Z+and+Datastreams%2FObservations%2FphenomenonTime+le+2019-08-07T05:59:59.999Z&$top=10&$expand=FeatureOfInterest,Datastream,Datastream%2FSensor,Datastream%2FObservedProperty
If it shows nothing it should mean that there were no data for that specific date range you selected. It can be very helpful for me if you can provide me more details on the date range you tried and the sensor ID.**

> * There are several unnecessary `=` in the exported data line. Is this on purpose?  If so, why? If not, please remove them.
> 
> ```
> "=""2019-08-05T18:04:12.952Z""","=""32.57""","=""Triangle Mtn - Langford""","=""Victoria""","=""{""coordinates"":[-123.514672,48.418927],""type"":""Point""}""","=""%""","=""Humidity""","=""SHT31 sensor"""
> ```

**No it was not on purpose, I have used vue-json-excel for generating exported file. I will take a look at that and try to remove unnecessary '='s.**
 
> 
> **Optional**
> 
> * Could put property description and unit next to observation result (result, property, unit)
> * Might make the preview table higher to disable the table inline scroll.  Therefore, the user can scroll down the page with only the page scroll, which is more easily.
> * Might add a link,  which is a go back to the sensor list link,  next to the Device name, like [Sensor List](previous_link) / Device 24EC64 / Export
> * Remove all selected properties from the dropdown list to reduce confusion, in other words, the dropdown list only display unselected properties
> * Disabled the ‘show preview’ button during the data exporting as well

**Thank you @kanluo for your valuable comments. I will review all of your comments and try to apply all the required improvements.**

kanluo commented:
I select the first sensor and choose the date-time range from Aug 1-2, but it displays today's data. 
![preview](https://user-images.githubusercontent.com/42560590/62581376-3365df00-b866-11e9-94ef-200c29c4278a.gif)

soroushojagh6633 commented:
> I select the first sensor and choose the date-time range from Aug 1-3, but it displays today's data.
> ![preview](https://user-images.githubusercontent.com/42560590/62581376-3365df00-b866-11e9-94ef-200c29c4278a.gif)
> I see it will send a new request when you change the date range and receive data, but the table doesn't display data correctly.

Thanks @kanluo, we edited the query and changed the filter section. It works fine now.

mahjahromi committed:
Updated download.vue and sta-request.js

Download.vue:
1- Added a link back to the sensor list page
2- Disabled the preview button while the data is exporting
3- Increased the height of the preview table so that the table inline scroll is disabled

sta-request.js:
1- FIxed the bug with querying data based on phenomenonTime

soroushojagh6633 commented:
> Overall, this looks good to me. Good job. I find some problems you could fix.
> 
> **Important**
> 
> * It didn’t display preview data correctly when you provide a given data range. It either displays the latest 10 data or nothing.
> * There are several unnecessary `=` in the exported data line. Is this on purpose?  If so, why? If not, please remove them.
> 
> ```
> "=""2019-08-05T18:04:12.952Z""","=""32.57""","=""Triangle Mtn - Langford""","=""Victoria""","=""{""coordinates"":[-123.514672,48.418927],""type"":""Point""}""","=""%""","=""Humidity""","=""SHT31 sensor"""
> ```
> 
> **Optional**
> 
> * Could put property description and unit next to observation result (result, property, unit)
> * Might make the preview table higher to disable the table inline scroll.  Therefore, the user can scroll down the page with only the page scroll, which is more easily.
> * Might add a link,  which is a go back to the sensor list link,  next to the Device name, like [Sensor List](previous_link) / Device 24EC64 / Export
> * Remove all selected properties from the dropdown list to reduce confusion, in other words, the dropdown list only display unselected properties
> * Disabled the ‘show preview’ button during the data exporting as well

Data columns were reordered based on your suggestion @kanluo

soroushojagh6633 commented:
> * Might make the preview table higher to disable the table inline scroll.  Therefore, the user can scroll down the page with only the page scroll, which is more easily.



> * ould put property description and unit next to ob
Optional suggestions:
1. Done
2. Done
3. Done
4. Done
5. Done

soroushojagh6633 commented:
> > Overall, this looks good to me. Good job. I find some problems you could fix.
> > **Important**
> > 
> > * It didn’t display preview data correctly when you provide a given data range. It either displays the latest 10 data or nothing.
> > * There are several unnecessary `=` in the exported data line. Is this on purpose?  If so, why? If not, please remove them.
> > 
> > ```
> > "=""2019-08-05T18:04:12.952Z""","=""32.57""","=""Triangle Mtn - Langford""","=""Victoria""","=""{""coordinates"":[-123.514672,48.418927],""type"":""Point""}""","=""%""","=""Humidity""","=""SHT31 sensor"""
> > ```
> > 
> > 
> > **Optional**
> > 
> > * Could put property description and unit next to observation result (result, property, unit)
> > * Might make the preview table higher to disable the table inline scroll.  Therefore, the user can scroll down the page with only the page scroll, which is more easily.
> > * Might add a link,  which is a go back to the sensor list link,  next to the Device name, like [Sensor List](previous_link) / Device 24EC64 / Export
> > * Remove all selected properties from the dropdown list to reduce confusion, in other words, the dropdown list only display unselected properties
> > * Disabled the ‘show preview’ button during the data exporting as well
> 
> Data columns were reordered based on your suggestion @kanluo
> The hight is set for the preview table
> The link is added before the device MacID
> selected properties are removed from the dropdown list
> 'Sow preview' button is disabled when user click on export button

kanluo commented:
1. I still have some issues with the preview function. 
<img width="1552" alt="Screen Shot 2019-08-08 at 12 11 05" src="https://user-images.githubusercontent.com/42560590/62727081-e7d24300-b9d5-11e9-8f86-9e6e669b3160.png">
Please test the function carefully.


2. > Data columns were reordered based on your suggestion

I don't see the data columns are reordered. 

3. > For the preview option the number of requested entities are limited to top 10, and that of for the download option is set to top 2000

Do you increase the number of preview data from 10 to 30?

soroushojagh6633 committed:
JsonExcelCustomized added

soroushojagh6633 committed:
Update sta-headers.js

soroushojagh6633 committed:
Updated download and sta-request

soroushojagh6633 committed:
Update staDataExporter.vue

soroushojagh6633 commented:
> 1. I still have some issues with the preview function.
> 
> <img alt="Screen Shot 2019-08-08 at 12 11 05" width="1552" src="https://user-images.githubusercontent.com/42560590/62727081-e7d24300-b9d5-11e9-8f86-9e6e669b3160.png">
> 
> Please test the function carefully.

@kanluo Please check the latest committed code

> 1. > Data columns were reordered based on your suggestion
> 
> I don't see the data columns are reordered. (result, unit, property)
> 
@kanluo Please check the latest committed code
> 1. > For the preview option the number of requested entities are limited to top 10, and that of for the download option is set to top 2000
> 
> Do you increase the number of preview data from 10 to 30?

No, it is still set to top 10

kanluo commented:
@soroushojagh6633 
> No, it is still set to top 10

I find it displays the top 30 records now
![Screen Shot 2019-08-08 at 15 25 45](https://user-images.githubusercontent.com/42560590/62738945-cfbbed00-b9f0-11e9-9e1f-af26f98091c1.png)
.

kanluo commented:
Just this, everything else is fine.
